File: totp.ts

//totp --> typescript
import base32 from 'hi-base32';
import { createHmac } from 'crypto'; //node crypto

 private generateHOTP(secret: string, counter: number) :string  {

    const len = 6;
    const buffer = Buffer.alloc(8);

    if (!secret || secret.length <= 0) {
      return this.leftpad("", len);    }    

    const decodedKey = base32.decode.asBytes(secret)
       
    for (let i = 0; i < 8; i++) {
      buffer[7 - i] = counter & 0xff;
      counter = counter >> 8;
    }

    const hmacObj = createHmac('sha1', Buffer.from(decodedKey));
    hmacObj.update(buffer);
    let hmac = hmacObj.digest();

    const hotp = this.dynamicTruncation(hmac);
    const code = hotp % (10 ** len);

    return this.leftpad(code.toString(10), len);
  }

  private dynamicTruncation(hmacVal: Buffer) {
    const offset = hmacVal[hmacVal.length - 1] & 0xf;

    return (
      ((hmacVal[offset] & 0x7f) << 24) |
      ((hmacVal[offset + 1] & 0xff) << 16) |
      ((hmacVal[offset + 2] & 0xff) << 8) |
      (hmacVal[offset + 3] & 0xff)

    );
  }

  private leftpad(str: string, len: number) : string {
    const pad = '0';
  if (len + 1 >= str.length) {
    str = Array(len + 1 - str.length).join(pad) + str;
  }
  return str;
}

  public TOTP(secret: string, window: number = 1) {
    const stepTime = 120;
    const counter = Math.floor(Date.now() / (stepTime * 1000));
    return this.generateHOTP(secret, counter+window)
  }
  
  
  

File: totp.cs
  //totp c# 
  
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Security.Cryptography;

namespace Utils.Cryptography
{
    public class Totp
    {
        private readonly long unixEpochTicks = 621355968000000000L;
        private readonly long ticksToSeconds = 10000000L;

        public bool Validate(string code, string secret, ValidateOpts opts)
        {
             var window = CalculateTimeStepFromTimestamp(DateTime.UtcNow, opts.TimePeriod);
         
            var counterBytes = new byte[8];
            long counter = window + opts.Skew;

            //for (var i = counterBytes.Length - 1; i >= 0; i--)
            //{
            //    counterBytes[i] = (byte)(counter & 0xff);
            //    counter >>= 8;
            //}

            for (var i = 0; i < 8; i++)
            {
                counterBytes[7 - i] = /*counter & 0xff;*/ (byte)(counter & 0xff);
                counter = counter >> 8;
            }

            opts.Algorithm.Key = Base32.FromBase32String(secret);
            byte[] hmacComputedHash = opts.Algorithm.ComputeHash(counterBytes);

            int gcode = Truncate(hmacComputedHash, opts.Digits);
            var scode = gcode.ToString().PadLeft(opts.Digits, '0');

            return string.Equals(code, scode);
        }


        private long CalculateTimeStepFromTimestamp(DateTime timestamp, int step)
        {
            var unixTimestamp = (timestamp.Ticks - unixEpochTicks) / ticksToSeconds;
            long window = unixTimestamp / step;
            return window;
        }

        private int DT(byte[] hmac_result)
        {
            var offset = hmac_result[hmac_result.Length - 1] & 0xf;
            var bin_code = (hmac_result[offset] & 0x7f) << 24
               | (hmac_result[offset + 1] & 0xff) << 16
               | (hmac_result[offset + 2] & 0xff) << 8
               | (hmac_result[offset + 3] & 0xff);

            return bin_code;
        }

        private int Truncate(byte[] hmac_result, int digits)
        {
            var Snum = DT(hmac_result);
            return Snum % (int)Math.Pow(10, digits);
        }
    }

    public class ValidateOpts
    {

        public int TimePeriod { get; private set; }
        public int Skew { get; private set; }
        public int Digits { get; private set; }
        public HMAC Algorithm { get; private set; }

        public ValidateOpts()
        {
            TimePeriod = 30;
            Skew = 0;
            Digits = 6;
            Algorithm = new HMACSHA1();
        }

        public ValidateOpts(int step, int skew, int digits, HMAC algo)
        {
            TimePeriod = step;
            Skew = skew;
            Digits = digits;
            Algorithm = algo;
        }
    }
}
